src/adapter/tracer/ptrace.rs

```diff
@@ -0,0 +1,453 @@
+///
+/// This struct is intended for use inside of a TracingAdapter.
+/// Operations must be recorded sequentially in chronological order.
+/// Recording out-of-order operations will lead to invalid state.
+#[derive(Debug, Clone)]
+pub struct Tracer {
+    pub calls: BTreeMap<FunctionCall, Summary>,
+
+    // When we measure the time of an iterator, we also measure the time spent
+    // evaluating its inputs. We must therefore subtract this time out when
+    // we record the time each operation takes.
+    last_input_duration: Option<Duration>,
+}
+
+impl Tracer {
+    /// Construct a new tracer.
+    pub fn new() -> Self {
+        Self {
+            calls: BTreeMap::new(),
+            last_input_duration: None,
+        }
+    }
+
+    /// Record an operation.
+    pub fn record_time(&mut self, call_id: &FunctionCall, duration: Duration) {
+        if let Some(summary) = self.calls.get_mut(call_id) {
+            summary.update(duration);
+        } else {
+            self.calls.insert(call_id.clone(), Summary::new(duration));
+        }
+    }
+
+    /// Set the duration of the last input.
+    pub fn record_last_input_duration(&mut self, duration: Duration) {
+        self.last_input_duration = Some(duration);
+    }
+
+    /// Get the duration of the last input. Panics if the duration is None.
+    pub fn get_last_input_duration(&self) -> Duration {
+        self.last_input_duration.unwrap()
+    }
+}
+
+impl Default for Tracer {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+#[allow(clippy::enum_variant_names)] // Names match the adapter functions 
+pub enum FunctionCall {
+    ResolveProperty(Vid, Arc<str>, Arc<str>), // vertex ID + type name + name of the property
+    ResolveNeighbors(Vid, Arc<str>, Eid),     // vertex ID + type name + edge ID
+    ResolveNeighborsInner(Vid, Arc<str>, Eid), // same as ResolveNeighbors
+    ResolveCoercion(Vid, Arc<str>, Arc<str>), // vertex ID + current type + coerced-to type
+}
+
+struct PerfSpanIter<I, T, F>
+where
+    I: Iterator<Item = T>,
+    F: Fn(T, Duration) -> T,
+{
+    inner: I,
+    post_action: F,
+}
+
+impl<I, T, F> Iterator for PerfSpanIter<I, T, F>
+where
+    I: Iterator<Item = T>,
+    F: Fn(T, Duration) -> T,
+{
+    type Item = T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let start = Instant::now();
+        let item = self.inner.next();
+        let time = start.elapsed();
+        match item {
+            Some(item) => Some((self.post_action)(item, time)),
+            None => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.inner.size_hint()
+    }
+}
+
+fn make_iter_with_perf_span<I, T, F>(inner: I, post_action: F) -> PerfSpanIter<I, T, F>
+where
+    I: Iterator<Item = T>,
+    F: Fn(T, Duration) -> T,
+{
+    PerfSpanIter { inner, post_action }
+}
+
+/// Traces an inner adapter. Constructed with new(). Each tracer is only valid
```

Comment:
When referring to Rust items in doc comments, always use rustdoc references like this:
```suggestion
/// Traces an inner adapter. Constructed with [`TracingAdapter::new()`]. Each tracer is only valid
```
Rustdoc will check the validity of the reference (i.e. ensure it isn't dangling) and then link to the item in the generated HTML documentation.

Make sure to apply the same to `finish()` below, and you may need to reflow the text to make the lines not be too long.